### PR TITLE
VGG Implementation

### PR DESCRIPTION
**Pull request by shubham1206agra**
Implemented VGG using the new ann-vtable structure as a Layer for more straightforward serialization and use.

### CODE REVIEW

File: models/vgg/vgg.hpp [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+#include <mlpack/methods/ann/init_rules/random_init.hpp>
+#include <mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp>
+
+namespace mlpack {
+namespace models {
+
+/**
+ * Definition of a VGG CNN.
+ * 
+ * NOTE: Note that output size will be 1x1xN. Here, N is number of classes.
```
Comment (rcurtin) [inline models/vgg/vgg.hpp:41]: Do you mean that we will return a `Cube` for each prediction?  It seems like we should be able to return a length-`N` vector for each input point.

File: models/vgg/vgg.hpp [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+ * 
+ * @tparam MatType Matrix representation to accept as input and use for
+ *    computation.
+ * @tparam VGGVersion Version of VGG.
+ * @tparam IsBatchNorm Whether to apply Batch Norm layer.
```
Comment (rcurtin) [inline models/vgg/vgg.hpp:46]: I might suggest either `UsesBatchNorm` or `IsBatchNormalized` (since `BatchNorm` is the name of a layer).  That's a small comment though.

File: models/vgg/vgg.hpp [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+  /**
+   * VGGType constructor intializes number of classes and weights.
+   *
+   * @param numClasses Optional number of classes to classify images into,
+   *     only to be specified if includeTop is  true.
+   * @param includeTop Must be set to true if weights are set.
+   */
+  VGGType(
+    const size_t numClasses,
+    const bool includeTop = true);
```
Comment (rcurtin) [inline models/vgg/vgg.hpp:68]:
```suggestion
  VGGType(const size_t numClasses,
          const bool includeTop = true);
```

File: models/vgg/vgg.hpp [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+{
+ public:
+  //! Create the VGGType layer.
+  VGGType();
+
+  /**
+   * VGGType constructor intializes number of classes and weights.
+   *
+   * @param numClasses Optional number of classes to classify images into,
+   *     only to be specified if includeTop is  true.
```
Comment (rcurtin) [inline models/vgg/vgg.hpp:63]: I don't think that I understand what `includeTop` does; can you clarify that in the documentation?

File: models/vgg/vgg.hpp [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+  //! Virtual destructor: delete all held layers.
+  virtual ~VGGType()
+  { /* Nothing to do here. */ }
+
+  //! Create a copy of the VGGType (this is safe for polymorphic use).
+  VGGType* Clone() const { return new VGGType(*this); }
+
+  /**
+   * Get Layers of the model.
```
Comment (rcurtin) [inline models/vgg/vgg.hpp:87]: I would write `Get the FFN object representing the network` just for clarity.

File: models/vgg/vgg.hpp [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    vgg->Add(this);
+    return vgg;
+  }
+
+  //! Serialize the VGGType.
+  template<typename Archive>
+  void serialize(Archive& ar, const uint32_t /* version */);
+
+ private:
+  void makeModel()
```
Comment (rcurtin) [inline models/vgg/vgg.hpp:109]:
```suggestion
  void MakeModel()
```
Just a stylistic convention. :)  Also I would move the implementation into `vgg_impl.hpp` to keep the header file shorter.

File: models/vgg/vgg.hpp [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+  {
+    std::map<size_t, std::vector<size_t>> const construct {
+      { 11, {64, 0, 128, 0, 256, 256, 0, 512, 512, 0, 512, 512, 0} },
+      { 13, {64, 64, 0, 128, 128, 0, 256, 256, 0, 512, 512, 0, 512, 512, 0} },
+      { 16, {64, 64, 0, 128, 128, 0, 256, 256, 256, 0, 512, 512, 512, 0, 512,
+          512, 512, 0} },
+      { 19, {64, 64, 0, 128, 128, 0, 256, 256, 256, 256, 0, 512, 512, 512, 512,
+          0, 512, 512, 512, 512, 0} }
```
Comment (rcurtin) [inline models/vgg/vgg.hpp:117]: Can you add a comment indicating what this represents?  I believe it is the network structure of each version (A, A-LRN, B, C, D, E) but I am not quite sure what the `std::vector<size_t>` associated with each model is.

Comment (rcurtin) [reply models/vgg/vgg.hpp:117]: I guess that `0` indicates a `MaxPooling` layer.  Anyway, it would be helpful to describe that in a comment.

File: models/vgg/vgg.hpp [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+namespace models {
+
+/**
+ * Definition of a VGG CNN.
+ * 
+ * NOTE: Note that output size will be 1x1xN. Here, N is number of classes.
+ * 
+ * @tparam MatType Matrix representation to accept as input and use for
+ *    computation.
+ * @tparam VGGVersion Version of VGG.
```
Comment (rcurtin) [inline models/vgg/vgg.hpp:45]: It would probably be a good idea to write which versions are valid here.  I guess it must be `11`, `13`, `16`, and `19`; but how would the user get `A-LRN` or `C`?  Is that done by setting the batch normalization parameter to true?

File: models/vgg/vgg.hpp [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    typename OutputLayerType = ann::CrossEntropyError,
+    typename InitializationRuleType = ann::RandomInitialization
+  >
+  ann::FFN<OutputLayerType, InitializationRuleType, MatType>* GetModel()
+  {
+    ann::FFN<OutputLayerType, InitializationRuleType, MatType>* vgg =
+        new ann::FFN<OutputLayerType, InitializationRuleType, MatType>();
+    vgg->Add(this);
+    return vgg;
+  }
```
Comment (rcurtin) [inline models/vgg/vgg.hpp:102]: Is the user supposed to call `GetModel()` and then use `Train()` and `Predict()`?  Or should we provide a `Train()` and `Predict()` function here that simply call the internal model?

(I am not 100% familiar with the models repository, so maybe all the models are like this, I am not sure.)

**Review by rcurtin — COMMENTED** [review]
Nice! The general implementation looks correct to me.  However I suspect that this is really, really painful to run with OpenBLAS instead of on a GPU. :)  Have you managed to run it at all?

It would be nice to add some simple tests too, so that we know that VGG compiles and runs.  Of course these would have to be very simple tests (like making sure that we can do a forward pass on one point, and a backwards pass on one point, and that the output is a reasonable size), so that the test does not take five years to run. :)

File: models/vgg/vgg.hpp (hunk 1 above)
Comment (shubham1206agra) [reply models/vgg/vgg.hpp:41]: I left that comment by mistake.

File: models/vgg/vgg.hpp (hunk 8 above)
Comment (shubham1206agra) [reply models/vgg/vgg.hpp:45]: Idts I have implemented `A-LRN`.

File: models/vgg/vgg.hpp (hunk 9 above)
Comment (shubham1206agra) [reply models/vgg/vgg.hpp:102]: No need as it is contained inside FFN generated by `GetModel()`.

File: models/vgg/vgg.hpp [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+class VGGType : public ann::MultiLayer<MatType>
+{
+ public:
+  //! Create the VGGType layer.
+  VGGType();
+
+  /**
+   * VGGType constructor intializes number of classes and weights.
+   *
+   * @param numClasses Optional number of classes to classify images into,
```
Comment (rcurtin) [inline models/vgg/vgg.hpp:60]: There's no default, so is this optional?  Maybe it makes sense to have `includeTop` before `numClasses`.

File: models/vgg/vgg.hpp (hunk 9 above)
Comment (rcurtin) [reply models/vgg/vgg.hpp:99]: Understood, thanks!

File: models/vgg/vgg.hpp [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/layer/layer_types.hpp>
+#include <mlpack/methods/ann/init_rules/random_init.hpp>
+#include <mlpack/methods/ann/loss_functions/binary_cross_entropy_loss.hpp>
+
+namespace mlpack {
+namespace models {
+
+/**
+ * Definition of a VGG CNN.
```
Comment (rcurtin) [inline models/vgg/vgg.hpp:39]: Actually it's probably better to put the paper reference here with the class, since users probably will look for documentation for the class and not necessarily the header file.

Do you think you can add a note on how `VGGVersion` relates to Table 1 in the paper?  i.e. `11` is equivalent to the `A` network, etc., etc.

**Review by rcurtin — COMMENTED** [review]
Everything is looking good to me here, I only have a few tiny change requests, and then I think it's ready to merge if you're done with it. :+1:

**Comment by rcurtin** [comment]
Ah, oops, I forgot, we also talked about adding some simple tests:

> It would be nice to add some simple tests too, so that we know that VGG compiles and runs. Of course these would have to be very simple tests (like making sure that we can do a forward pass on one point, and a backwards pass on one point, and that the output is a reasonable size), so that the test does not take five years to run. :)

File: models/vgg/vgg.hpp (hunk 10 above)
Comment (shubham1206agra) [reply models/vgg/vgg.hpp:60]: Writing error :)

File: models/vgg/vgg.hpp (hunk 11 above)
Comment (shubham1206agra) [reply models/vgg/vgg.hpp:39]: Idts they will look for paper here.

Comment (rcurtin) [reply models/vgg/vgg.hpp:39]: It depends on the tool they are using; if they are just looking through files, I agree, but if they are using some IDE that points them to the class documentation, they will end up looking at this particular comment.  It is a minor issue, so if you feel strongly about it, it's okay to leave it how it is, but in my opinion it would be better placed here. :+1:

File: tests/vgg_tests.cpp [hunk 12]
```diff
@@ -0,0 +1,507 @@
...
+template<typename LayerType>
+void ModelSerializationTest(LayerType& layer)
+{
+  arma::mat input(224 * 224 * 3, 10, arma::fill::randu);
+  arma::mat output(1000, 10, arma::fill::randu);
+
+  ann::FFN<> model;
+  model.Add<LayerType>(layer);
+
+	model.InputDimensions() = std::vector<size_t>({224, 224, 3});
```
Comment (rcurtin) [inline tests/vgg_tests.cpp:84]: Looks like there are some style issues, with tab characters appearing here and in the later tests.  Can you fix those?

File: tests/vgg_tests.cpp [hunk 13]
```diff
@@ -0,0 +1,507 @@
...
+
+  ens::StandardSGD opt(0.1, 1, 5, -100, false);
```
Comment (rcurtin) [inline tests/vgg_tests.cpp:86]: Might be nice to add a comment that this only takes one pass over the data (so it does not take super long).

**Review by rcurtin — APPROVED** [review]
Nice, thanks for adding the tests.  I don't think I have any more comments; if you can fix the style issues, I'll go ahead and merge. :+1:

**Comment by rcurtin** [comment]
Awesome work @shubham1206agra! :+1: Just two more to merge :)
